File: JAVA + DSA/Codes/DSA/Linked List/Doubly Linked List/DLL.java
import java.util.Scanner;

class Node {
    int data;
    Node next;
    Node prev;

    Node(int d) {
        data = d;
        next = null;
        prev = null;
    }
}

class DLL {
    Node head;
    int count;

    DLL() {
        head = null;
        count = 0;
    }

    // Insert at start
    void insertAtStart(int data) {
        Node newNode = new Node(data);
        count++;

        if (head == null) {
            head = newNode;
            return;
        }

        head.prev = newNode;
        newNode.next = head;
        head = newNode;
    }

    // Insert at end
    void insertAtEnd(int data) {
        Node newNode = new Node(data);
        count++;

        if (head == null) {
            head = newNode;
            return;
        }

        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }

        temp.next = newNode;
        newNode.prev = temp;
    }

    // Insert by position (currently not implemented)
    void insertByPosition(int data) {
        System.out.println("Insert by position not implemented yet.");
    }

    // Delete at start
    void deleteAtStart() {
        if (head == null) {
            System.out.println("LIST IS EMPTY");
            return;
        }

        if (head.next == null) {
            deleteSingleNode();
            return;
        }

        Node temp = head;
        head = head.next;
        head.prev = null;
        count--;
    }

    // Delete at end
    void deleteAtEnd() {
        if (head == null) {
            System.out.println("LIST EMPTY");
            return;
        }

        if (head.next == null) {
            deleteSingleNode();
            return;
        }

        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }

        temp.prev.next = null;
        count--;
    }

    // Delete single node
    void deleteSingleNode() {
        count--;
        head = null;
    }

    // Traverse
    void traverse() {
        if (head == null) {
            System.out.println("LIST IS EMPTY");
            return;
        }

        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }
}

public class DoublyLinkedList {
    public static void main(String[] args) {
        DLL list = new DLL();
        Scanner sc = new Scanner(System.in);
        boolean isExit = false;

        while (!isExit) {
            System.out.println("\n1. Insert At Start");
            System.out.println("2. Insert At End");
            System.out.println("3. Traverse");
            System.out.println("5. Insert By Position");
            System.out.println("4. Exit");
            System.out.println("6. Delete at Start");
            System.out.println("7. Delete at End");

            System.out.print("\nEnter the choice: ");
            int ch = sc.nextInt();

            switch (ch) {
                case 1:
                    System.out.print("Insert the data: ");
                    int data1 = sc.nextInt();
                    list.insertAtStart(data1);
                    break;

                case 2:
                    System.out.print("Insert the data: ");
                    int data2 = sc.nextInt();
                    list.insertAtEnd(data2);
                    break;

                case 3:
                    list.traverse();
                    break;

                case 4:
                    isExit = true;
                    break;

                case 5:
                    System.out.print("Insert the data: ");
                    int data5 = sc.nextInt();
                    list.insertByPosition(data5);
                    break;

                case 6:
                    list.deleteAtStart();
                    break;

                case 7:
                    list.deleteAtEnd();
                    break;

                default:
                    System.out.println("Invalid choice. Try again.");
            }
        }

        sc.close();
        System.out.println("\nProgram exited.");
    }
}

File: CPP + DSA/Codes/DSA/Linked List/Doubly Linked List/DLL.cpp
#include <bits/stdc++.h>
using namespace std;


class Node{
public:
    int data;
    Node *next;
    Node *prev;

    Node(int d){
        data = d;
        next = NULL;
        prev = NULL;
    }

};

class DLL{
public:
    Node *head;
    int count;

    DLL(){
        head = NULL;
        count = 0;
    }   

    // ******** INSERTION OPERATION ************

    // insert at start
    void insertAtStart(int data){

        Node *newNode = new Node(data);
        count++;

        // if list is empty, then node itself is a first node
        if (head == NULL){
            head = newNode;
            return;
        }

        // 1 2 3 4

        // newNode = 10

        // 10 1 2 3 4

        head -> prev = newNode;
        newNode -> next = head;
        head = newNode;
    }

    // insert at end
    void insertAtEnd(int data){

        Node *newNode = new Node(data);
        count++;

         // if list is empty, then node itself is a first node
        if (head == NULL){
            head = newNode;
            return;
        }

        Node *temp = head;

        while(temp -> next != NULL){
            temp = temp -> next;
        }

        temp -> next = newNode;
        newNode -> prev = temp;

    }

    // insert by position
    void insertByPosition(int data){
        
    }

    // ******** DELETION OPERATIONS ************

    // delete at start
    void deleteAtStart(){

        // if list is empty
        if (head == NULL){
            cout << "LIST IS EMPTY\n";
            return;
        }

        // if list contains only one node
        if (head -> next == NULL){
            deleteSingleNode();
            return;
        }

        Node *temp = head;
        head = head -> next;
        head -> prev = NULL;
        delete(temp);
        count--;
    }

    // function to delete single node
    void deleteSingleNode(){
        count--;
        Node *temp = head;
        head = NULL;
        delete(temp);
    }

    // delete at end
    void deleteAtEnd(){

        // if list is empty
        if (head == NULL){
            cout << "LIST EMPTY\n";
            return;
        }

        // if list contains only one node
        if (head -> next == NULL){
            deleteSingleNode();
            return;
        }

        Node *temp = head;

        while(temp -> next != NULL){
            temp = temp -> next;
        }

        temp -> prev -> next = NULL;
        delete(temp);
        count--;
    }


    // traversal
    void traverse(){

        //if list is empty
        if (head == NULL){
            cout << "LIST IS EMPTY\n";
            return;
        }

        Node *temp = head;

        while(temp != NULL){
            cout << temp -> data << " ";
            temp = temp -> next;
        }

        cout << "\n";
    }
};



int main()
{

    DLL list;
    bool isExit = false;

    while(1){

        cout << "1. Insert At Start\n";
        cout << "2. Insert At End\n";
        cout << "3. Traverse\n";
        cout << "5. Insert By position\n";
        cout << "4. Exit\n";
        cout << "6. Delete at start\n";
        cout << "7. Delete at End\n";

        int ch;
        cout << "\nEnter the choice: ";
        cin >> ch;

        switch(ch){

            case 1:{
                int data;
                cout << "\nInsert the data which you want\n";
                cin >> data;
                list.insertAtStart(data);
                break;
            }

             case 2:{
                int data;
                cout << "\nInsert the data which you want\n";
                cin >> data;
                list.insertAtEnd(data);
                break;
            }

             case 3:{
                list.traverse();
                break;
            }

            case 4:{
                isExit = true;
                break;
            }

            case 5:{
                int data;
                cout << "\nInsert the data which you want\n";
                cin >> data;
                list.insertByPosition(data);
                break;
            }

            case 6:{
                list.deleteAtStart();
                break;
            }

            case 7:{
                list.deleteAtEnd();
                break;
            }

        }

        if (isExit == true){
            break;
        }

    }
  
    cout << "\n";

    return 0;
}